shelley/chain-and-ledger/executable-spec/src/LedgerState.hs

```diff
@@ -394,15 +396,68 @@ validInputs tx u =
     else Invalid [BadInputs]
 
 -- |Implementation of abstract transaction size
-txsize :: TxBody crypto-> Integer
-txsize = toEnum . length . show
+-- TODO we still need to account for the witnesses, see github issue #812
+txsize :: forall crypto . (Crypto crypto) => TxBody crypto-> Integer
+txsize (TxBody ins outs cs ws _ _ (Update (PPUpdate ppup) (AVUpdate avup))) =
+  iSize + oSize + cSize + wSize + feeSize + ttlSize + uSize
+  where
+    hl = toInteger $ byteCount (Proxy :: Proxy (HASH crypto))
+    hashObj = 2 + hl
+    uint = 5
+    arrayPrefix = 2
+    smallArray = 1
+    mapPrefix = 2
+    label = 1
+    cborTag = 2
+    address = 2 * hashObj
+    credential = label + hashObj
+    unitInterval = cborTag + smallArray + uint + uint
+    feeSize = label + uint
+    ttlSize = label + uint
+    iSize = label + cborTag + arrayPrefix
+            + (toInteger $ length ins) * (smallArray + uint + hashObj)
+    oSize = label + arrayPrefix
+            + (toInteger $ length outs) * (smallArray + uint + address)
+
+    numPoolOwners = toInteger . length . _poolOwners
+    pparams pps = cborTag + arrayPrefix + (numPoolOwners pps)*(hashObj) -- pool owners
+                + uint -- cost
+                + unitInterval -- margin
+                + uint -- pledge
+                + hashObj -- operator
+                + hashObj -- vrf keyhash
+                + credential -- reward account
+    certSize (DCertDeleg (RegKey _)) = smallArray + label + hashObj
+    certSize (DCertDeleg (DeRegKey _)) = smallArray + label + hashObj
+    certSize (DCertDeleg (Delegate _ )) = smallArray + label + 2 * hashObj
+    certSize (DCertPool (RegPool pps)) = smallArray + label + hashObj + pparams pps
+    certSize (DCertPool (RetirePool _ _)) = smallArray + label + uint + hashObj
+    certSize (DCertGenesis _) = smallArray + label + 2 * hashObj
+    certSize (DCertMir (MIRCert m)) = smallArray + label + mapPrefix
+                                    + (toInteger $ length m)*(uint + hashObj)
+    cSize = sum $ fmap certSize cs
+    wSize = label + mapPrefix + (toInteger $ length ws) * (uint + credential)
+
+    params = uint + uint + uint + uint + uint + uint
+               + unitInterval + unitInterval + uint + unitInterval
+               + unitInterval + uint + uint + unitInterval
+               + unitInterval + unitInterval + unitInterval + unitInterval
+               + uint + (arrayPrefix + uint + uint + uint)
```

Comment:
The protocol version is an array of size 3, so the prefix is only 1 byte instead of 2.